# Rosalind

Repository is for bioinformatics problems from <http://www.rosalind.info> and the textbook, "Bioinformatics Algorithms - An Active Learning Approach" by Phillip Compeau &amp; Pavel Pevzner.

## Completed "Bioinformatics Stronghold" Problems

ID   | Title
-----|------
DNA  | Counting DNA Nucleotides  
RNA  | Transcribing DNA into RNA  
REVC | Complementing a Strand of DNA  
FIB  | Rabbits and Recurrence Relations  
GC   | Computing GC Content  
HAMM | Counting Point Mutations  
IPRB | Mendel's First Law  
PROT | Translating RNA into Protein  
SUBS | Finding a Motif in DNA  
CONS | Consensus and Profile  
FIBD | Mortal Fibonacci Rabbits  
GRPH | Overlap Graphs  
IEV  | Calculating Expected Offspring  
LCSM | Finding a Shared Motif  
LIA  | Independent Alleles  
MPRT | Finding a Protein Motif  
MRNA | Inferring mRNA from Protein  
ORF  | Open Reading Frames  
PERM | Enumerating Gene Orders  
PRTM | Calculating Protein Mass  
REVP | Locating Restriction Sites  
SPLC | RNA Splicing  
LEXF | Enumerating k-mers Lexicographically  
LGIS | Longest Increasing Subsequence  
LONG | Genome Assembly as Shortest Superstring  
PMCH | Perfect Matchings and RNA Secondary Structures  
PPER | Partial Permutations  
PROB | Introduction to Random Strings  
SIGN | Enumerating Oriented Gene Orderings  
SSEQ | Finding a Spliced Motif  
TRAN | Transitions and Transversions  
TREE | Completing a Tree  
CAT  | Catalan Numbers and RNA Secondary Structures  
CORR | Error Correction in Reads  
INOD | Counting Phylogenetic Ancestors  
KMER | k-Mer Composition  
KMP  | Speeding Up Motif Finding  
LCSQ | Finding a Shared Spliced Motif  
LEXV | Ordering Strings of Varying Length Lexicographically  
MMCH | Maximum Matchings and RNA Secondary Structures  
PDST | Creating a Distance Matrix  
REAR | Reversal Distance  
RSTR | Matching Random Motifs  
SSET | Counting Subsets  
ASPC | Introduction to Alternative Splicing  
EDIT | Edit Distance  
EVAL | Expected Number of Restriction Sites  
MOTZ | Motzkin Numbers and RNA Secondary Structures  
NWCK | Distances in Trees  
SCSP | Interleaving Two Motifs  
SETO | Introduction to Set Operations  
SORT | Sorting by Reversals  
SPEC | Inferring Protein from Spectrum  
TRIE | Introduction to Pattern Matching  
CONV | Comparing Spectra with the Spectral Convolution  
CTBL | Creating a Character Table  
DBRU | Constructing a De Bruijn Graph  
EDTA | Edit Distance Alignment  
FULL | Inferring Peptide from Full Spectrum  
INDC | Independent Segregation of Chromosomes  
ITWV | Finding Disjoint Motifs in a Gene  
LREP | Finding the Longest Multiple Repeat  
NKEW | Newick Format with Edge Weights  
RNAS | Wobble Bonding and RNA Secondary Structures  
AFRQ | Counting Disease Carriers  
CTEA | Counting Optimal Alignments  
GLOB | Global Alignment with Scoring Matrix  
SGRA | Using the Spectrum Graph to Infer Peptides  
GCON | Global Alignment with Constant Gap Penalty  
LOCA | Local Alignment with Scoring Matrix  
MGAP | Maximizing the Gap Symbols of an Optimal Alignment  
MULT | Multiple Alignment  
SEXL | Sex-Linked Inheritance  
GAFF | Global Alignment with Scoring Matrix and Affine Gap Penalty
OAP  | Overlap Alignment
SIMS | Finding a Motif with Modifications
SMGB | Semiglobal Alignment
LAFF | Local Alignment with Affine Gap Penalty
OSYM | Isolating Symbols in Alignments
MEND | Inferring Genotype from a Pedigree
CSTR | Creating a Character Table from Genetic Strings
WFMD | The Wright-Fisher Model of Genetic Drift
EBIN | Wright-Fisher's Expected Behavior
FOUN | The Founder Effect and Genetic Drift
CUNR | Counting Unrooted Binary Trees
ROOT | Counting Rooted Binary Trees
PCOV | Genome Assembly with Perfect Coverage
ASMQ | Assessing Assembly Quality with N50 and N75
GASM | Genome Assembly Using Reads

## Unfinished "Bioinformatics Stronghold" Problems

ID   | Title
-----|------
PRSM | Matching a Spectrum to a Protein
QRT  | Quartets
SUFF | Encoding Suffix Trees
KSIM | Finding All Similar Motifs
CHBP | Character-Based Phylogeny
CNTQ | Counting Quartets
EUBT | Enumerating Unrooted Binary Trees
LING | Linguistic Complexity of a Genome
MREP | Identifying Maximal Repeats
PDPL | Creating a Restriction Map
SPTD | Phylogeny Comparison with Split Distance
ALPH | Alignment-Based Phylogeny
CSET | Fixing an Inconsistent Character Set
GREP | Genome Assembly with Perfect Coverage and Repeats
QRTD | Quartet Distance
RSUB | Identifying Reversing Substitutions
